File: tutorials/mpi/examples/mpi_prime.f
C ****************************************************************************
C FILE: mpi_prime.f
C DESCRIPTION:
C   Generates prime numbers.  All tasks distribute the work evenly, taking
C   every nth number, where n is the stride computed as:  (rank *2) + 1
C   so that even numbers are automatically skipped.  The method of using
C   stride is preferred over contiguous blocks of numbers, since numbers
C   in the higher range require more work to compute and may result in
C   load imbalance.  This program demonstrates embarrassing parallelism.
C   Collective communications calls are used to reduce the only two data
C   elements requiring communications: the number of primes found and
C   the largest prime.
C AUTHOR: Blaise Barney 11/25/95 - adapted from version contributed by 
C   Richard Ng & Wong Sze Cheong during MHPCC Singapore Workshop (8/22/95).
C LAST REVISED: 04/02/05
C ****************************************************************************
C Explanation of constants and variables
C   LIMIT                  = Increase this to find more primes 
C   FIRST                  = Rank of first task 
C   ntasks                 = total number of tasks in partition 
C   rank                   = task identifier 
C   n                      = loop variable 
C   pc                     = prime counter
C   pcsum                  = number of primes found by all tasks 
C   foundone               = most recent prime found 
C   maxprime               = largest prime found 
C   mystart                = where to start calculating 
C   stride                 = calculate every nth number

      program prime
      include 'mpif.h'

      integer LIMIT, FIRST
      parameter(LIMIT=2500000)        
      parameter(FIRST=0)

      integer ntasks, rank, ierr, n, pc, pcsum, foundone, maxprime,
     &        mystart, stride
      double precision start_time, end_time
      logical result

      call MPI_INIT(ierr)
      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
      call MPI_COMM_SIZE(MPI_COMM_WORLD, ntasks, ierr)
      if ((mod(ntasks,2).ne.0) .or. (mod(LIMIT,ntasks).ne.0)) then
         print *, mod(ntasks,2), mod(LIMIT,ntasks)
         print *,'Sorry -this exercise requires an even number of '
         print *,'processors evenly divisible into ',LIMIT,'.' 
         print *,'Try 4 or 8.'
         call MPI_FINALIZE(ierr)
         stop
      endif

C     Initializations: mystart must be odd number. stride is multiplied
C     by 2 to skip over even numbers.  
      start_time = MPI_WTIME()
      mystart = (rank*2) + 1
      stride = ntasks*2
      pc = 0
      foundone = 0

C ------------------- task with rank 0 does this part ------------------
      if (rank .eq. FIRST) then
         print *,'Using',ntasks,'tasks to scan',LIMIT,'numbers'
C        Assume first four primes are counted here
         pc = 4
         do n=mystart,LIMIT,stride
            call isprime(n,result)
            if (result .eqv. .true.) then
               pc = pc + 1
               foundone = n
C              ***** Optional: print each prime as it is found
C              print *, foundone
C              *****
            endif
         enddo
      call MPI_Reduce(pc,pcsum,1,MPI_INTEGER,MPI_SUM,FIRST,
     &                  MPI_COMM_WORLD,ierr)
      call MPI_Reduce(foundone,maxprime,1,MPI_INTEGER,MPI_MAX,
     &                 FIRST,MPI_COMM_WORLD,ierr)
      end_time=MPI_WTIME()
      print *,'Done. Largest prime is ',maxprime,' Total primes ',pcsum
      print *,'Wallclock time elapsed: ' ,end_time-start_time
      endif

C ------------------- all other tasks do this part ---------------------
      if (rank .gt. FIRST) then
         do n=mystart,LIMIT,stride
            call isprime(n,result)
            if (result .eqv. .true.) then
               pc = pc + 1
               foundone = n
C              ***** Optional: print each prime as it is found
C              print *, foundone
C              *****
            endif
         enddo
      call MPI_Reduce(pc,pcsum,1,MPI_INTEGER,MPI_SUM,FIRST,
     &                  MPI_COMM_WORLD,ierr)
      call MPI_Reduce(foundone,maxprime,1,MPI_INTEGER,MPI_MAX,
     &                 FIRST,MPI_COMM_WORLD,ierr)
      endif


      call MPI_FINALIZE(ierr)
      end


      subroutine isprime (n,result)
      integer n
      logical result
      integer i, squareroot
      real*4 realn
      if (n .gt. 10) then
         realn = real(n)
         squareroot = int (sqrt(realn))
         do i=3,squareroot,2
           if (mod(n,i) .eq. 0) then
              result = .false.
              return
           endif
         enddo
         result = .true.
         return 
C Assume first four primes are counted elsewhere. Forget everything else
      else 
        result = .false.
        return
      endif
      end

File: mpi/examples/mpi_prime.c
/******************************************************************************
* FILE: mpi_prime.c
* DESCRIPTION:
*   Generates prime numbers.  All tasks distribute the work evenly, taking
*   every nth number, where n is the stride computed as:  (rank *2) + 1
*   so that even numbers are automatically skipped.  The method of using
*   stride is preferred over contiguous blocks of numbers, since numbers
*   in the higher range require more work to compute and may result in 
*   load imbalance.  This program demonstrates embarrassing parallelism.
*   Collective communications calls are used to reduce the only two data
*   elements requiring communications: the number of primes found and
*   the largest prime.
* AUTHOR: Blaise Barney 11/25/95 - adapted from version contributed by 
*   Richard Ng &  Wong Sze Cheong during MHPCC Singapore Workshop (8/22/95).
* LAST REVISED: 04/13/05
******************************************************************************/
#include "mpi.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define LIMIT     2500000     /* Increase this to find more primes */
#define FIRST     0           /* Rank of first task */

int isprime(int n) {
int i,squareroot;
if (n>10) {
   squareroot = (int) sqrt(n);
   for (i=3; i<=squareroot; i=i+2)
      if ((n%i)==0)
         return 0;
   return 1;
   }
/* Assume first four primes are counted elsewhere. Forget everything else */
else
   return 0;
}


int main (int argc, char *argv[])
{
int   ntasks,               /* total number of tasks in partition */
      rank,                 /* task identifier */
      n,                    /* loop variable */
      pc,                   /* prime counter */
      pcsum,                /* number of primes found by all tasks */
      foundone,             /* most recent prime found */
      maxprime,             /* largest prime found */
      mystart,              /* where to start calculating */
      stride;               /* calculate every nth number */

double start_time,end_time;
	
MPI_Init(&argc,&argv);
MPI_Comm_rank(MPI_COMM_WORLD,&rank);
MPI_Comm_size(MPI_COMM_WORLD,&ntasks);
if (((ntasks%2) !=0) || ((LIMIT%ntasks) !=0)) {
   printf("Sorry - this exercise requires an even number of tasks.\n");
   printf("evenly divisible into %d.  Try 4 or 8.\n",LIMIT);
   MPI_Finalize();
   exit(0);
   }

start_time = MPI_Wtime();   /* Initialize start time */
mystart = (rank*2)+1;       /* Find my starting point - must be odd number */
stride = ntasks*2;          /* Determine stride, skipping even numbers */
pc=0;                       /* Initialize prime counter */
foundone = 0;               /* Initialize */

/******************** task with rank 0 does this part ********************/
if (rank == FIRST) {
   printf("Using %d tasks to scan %d numbers\n",ntasks,LIMIT);
   pc = 4;                  /* Assume first four primes are counted here */
   for (n=mystart; n<=LIMIT; n=n+stride) {
      if (isprime(n)) {
         pc++;
         foundone = n;
         /***** Optional: print each prime as it is found
         printf("%d\n",foundone);
         *****/
         }
      }
   MPI_Reduce(&pc,&pcsum,1,MPI_INT,MPI_SUM,FIRST,MPI_COMM_WORLD);
   MPI_Reduce(&foundone,&maxprime,1,MPI_INT,MPI_MAX,FIRST,MPI_COMM_WORLD);
   end_time=MPI_Wtime();
   printf("Done. Largest prime is %d Total primes %d\n",maxprime,pcsum);
   printf("Wallclock time elapsed: %.2lf seconds\n",end_time-start_time);
   }


/******************** all other tasks do this part ***********************/
if (rank > FIRST) {
   for (n=mystart; n<=LIMIT; n=n+stride) {
      if (isprime(n)) {
         pc++;
         foundone = n;
         /***** Optional: print each prime as it is found
         printf("%d\n",foundone);
         *****/
         }
      }
   MPI_Reduce(&pc,&pcsum,1,MPI_INT,MPI_SUM,FIRST,MPI_COMM_WORLD);
   MPI_Reduce(&foundone,&maxprime,1,MPI_INT,MPI_MAX,FIRST,MPI_COMM_WORLD);
   }

MPI_Finalize();
}
